src/android/android-uvc.cpp

```diff
@@ -0,0 +1,598 @@
+
+                        _power_state = D0;
+                        return;
+                    }
+                }
+
+                throw std::runtime_error("Device not found!");
+            }
+            if (state == D3 && _power_state == D0) {
+                _device.reset();
+                _power_state = D3;
+            }
+        }
+
+        std::vector<stream_profile> android_uvc_device::get_profiles() const {
+            std::vector<stream_profile> results;
+            int i = 0;
+
+            check_connection();
+
+            if (get_power_state() != D0) {
+                throw std::runtime_error("Device must be powered to query supported profiles!");
+            }
+
+            // Return linked list of uvc_format_t of all available formats inside device
+            uvc_format_t *formats = NULL;
+            usbhost_get_available_formats_all(_device.get(), &formats);
+
+            uvc_format_t *curFormat = formats;
+            while (curFormat != NULL) {
+                stream_profile sp;
+                sp.width = curFormat->width;
+                sp.height = curFormat->height;
+                sp.fps = curFormat->fps;
+                sp.format = curFormat->fourcc;
+                results.push_back(sp);
+                curFormat = curFormat->next;
+            }
+
+            usbhost_free_formats(formats);
+
+            return results;
+        }
+
+
+        //TODO: fix this
+        android_uvc_device::android_uvc_device(const uvc_device_info &info,
+                                               std::shared_ptr<const android_backend> backend)
+                : _streamIndex(MAX_PINS), _info(info), _backend(std::move(backend)),
+                  _location("")
+        {
+            if (!is_connected(info)) {
+                throw std::runtime_error("Camera not connected!");
+            }
+        }
+
+        android_uvc_device::~android_uvc_device() {
+            try {
+                android_uvc_device::set_power_state(D3);
+            }
+            catch (...) {
+                // TODO: Log
+            }
+        }
+
+        // profile config- set control
+        void android_uvc_device::probe_and_commit(stream_profile profile, frame_callback callback,
+                                                  int /*buffers*/) {
+            if (_streaming)
+                throw std::runtime_error("Device is already streaming!");
+
+            _profiles.push_back(profile);
+            _frame_callbacks.push_back(callback);
+        }
+
+        /* callback context send for each frame for its specific profile */
+        struct callback_context {
+            frame_callback _callback;
+            stream_profile _profile;
+            android_uvc_device *_this;
+        };
+
+        // receive the original callback and pass it to the right device. this is registered by libUVC.
+        static void internal_winusb_uvc_callback(frame_object *frame, void *ptr) {
```

Comment:
"WinUSB" in _my_ Android? It's more likely than you think ;)

Reply:
Gone in 3cc8826fa